codegenerator/cli/templates/dynamic/codegen/src/Types.res.hbs

```diff
@@ -234,225 +252,35 @@ module HandlerTypes = {
-          ~logger=Logging.createChildFrom(~logger, ~params=eventNamespace),
-          ~msg="Duplicate registration of event handlers not allowed",
-        )
-      }
-
-      switch value {
-        | {wildcard: ?None, eventFilters: ?None, preRegisterDynamicContracts: ?None} => ()
-        | {?wildcard, ?eventFilters, ?preRegisterDynamicContracts} =>
-        t->setEventOptions(
-          EventOptions.make(
-            ~isWildcard=wildcard->Option.getWithDefault(false),
-            ~eventFilters,
-            ~preRegisterDynamicContracts=preRegisterDynamicContracts->Option.getWithDefault(false),
-          ),
-          ~logger
-        )
-      }
-    }
-
-    let setContractRegister = (
-      t: t,
-      value,
-      ~eventOptions,
-      ~logger=Logging.getLogger(),
-    ) => {
-      switch t.contractRegister {
-      | None => t.contractRegister = Some(value->(Utils.magic: Internal.genericContractRegister<Internal.genericContractRegisterArgs<'event, 'context>> => Internal.contractRegister))
-      | Some(_) =>
-        let eventNamespace = {contractName: t.contractName, eventName: t.eventName}
-        DuplicateEventRegistration(eventNamespace)->ErrorHandling.mkLogAndRaise(
-          ~logger=Logging.createChildFrom(~logger, ~params=eventNamespace),
-          ~msg="Duplicate contractRegister handlers not allowed",
-        )
-      }
-      switch eventOptions {
-      | Some(eventOptions) => t->setEventOptions(eventOptions, ~logger)
-      | None => ()
-      }
-    }
-  }
+  type eventConfig<'eventFilters> = Internal.eventOptions<'eventFilters>
 }
 
 module type Event = {
   type event
 
-  type loader<'loaderReturn> = Internal.genericLoader<
-    Internal.genericLoaderArgs<event, loaderContext>,
-    'loaderReturn,
-  >
-  type handler<'loaderReturn> = Internal.genericHandler<
-    Internal.genericHandlerArgs<event, handlerContext, 'loaderReturn>,
-  >
-  type contractRegister = Internal.genericContractRegister<
-    Internal.genericContractRegisterArgs<event, contractRegistrations>,
-  >
-
-  let handlerRegister: HandlerTypes.Register.t
+  let handlerRegister: EventRegister.t
 
   type eventFilters
 }
 
-let makeEventOptions = (
-  type eventFilters,
-  eventConfig: option<HandlerTypes.eventConfig<eventFilters>>,
-) => {
-  open Belt
-  eventConfig->Option.map(({?wildcard, ?eventFilters, ?preRegisterDynamicContracts}) =>
-    HandlerTypes.EventOptions.make(
-      ~isWildcard=wildcard->Option.getWithDefault(false),
-      ~eventFilters,
-      ~preRegisterDynamicContracts=preRegisterDynamicContracts->Option.getWithDefault(false),
-    )
-  )
-}
-
 @genType.import(("./bindings/OpaqueTypes.ts", "HandlerWithOptions"))
 type fnWithEventConfig<'fn, 'eventConfig> = ('fn, ~eventConfig: 'eventConfig=?) => unit
 
+{{#if preload_handlers}}
+type handlerWithOptions<'eventArgs, 'eventFilters> = fnWithEventConfig<
+  Internal.genericHandler<'eventArgs>,
+  HandlerTypes.eventConfig<'eventFilters>,
+>
+{{else}}
 @genType
```

Comment:
Note: I decided not to include @genType for the new `preload_handlers` option. If somebody needs the types, we should make them importable from the `envio` package instead.